File: CodeBattleJava/src/main/java/ru/codebattle/client/api/BoardPoint.java
package ru.codebattle.client.api;

import lombok.AllArgsConstructor;
import lombok.Data;

import java.util.Objects;

@AllArgsConstructor
@Data
public class BoardPoint {
    private int x;
    private int y;

    /**
     * Checks is current point on board or out of range.
     *
     * @param boardSize Board size to compare
     */
    public boolean isOutOfBoard(int boardSize) {
        return x >= boardSize || y >= boardSize || x < 0 || y < 0;
    }

    /**
     * Returns new BoardPoint object shifted left to "delta" points
     */
    public BoardPoint shiftLeft(int delta) {
        return new BoardPoint(x - delta, y);
    }

    /**
     * Returns new BoardPoint object shifted left to 1 point
     */
    public BoardPoint shiftLeft() {
        return shiftLeft(1);
    }

    /**
     * Returns new BoardPoint object shifted right to "delta" points
     */
    public BoardPoint shiftRight(int delta) {
        return new BoardPoint(x + delta, y);
    }

    /**
     * Returns new BoardPoint object shifted right to 1 point
     */
    public BoardPoint shiftRight() {
        return shiftRight(1);
    }

    /**
     * Returns new BoardPoint object shifted top "delta" points
     */
    public BoardPoint shiftTop(int delta) {
        return new BoardPoint(x, y - delta);
    }

    /**
     * Returns new BoardPoint object shifted top 1 point
     */
    public BoardPoint shiftTop() {
        return shiftTop(1);
    }

    /**
     * Returns new BoardPoint object shifted bottom "delta" points
     */
    public BoardPoint shiftBottom(int delta) {
        return new BoardPoint(x, y + delta);
    }

    /**
     * Returns new BoardPoint object shifted bottom 1 point
     */
    public BoardPoint shiftBottom() {
        return shiftBottom(1);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        BoardPoint that = (BoardPoint) o;
        return x == that.x &&
                y == that.y;
    }

    public boolean notEquals(Object o) {
        return !equals(o);
    }

    @Override
    public int hashCode() {
        return Objects.hash(x, y);
    }

    @Override
    public String toString() {
        return String.format("[%s,%s]", x, y);
    }
}

File: CodeBattleCPP/CodeBattleCppLibrary/BoardPoint.cpp
#include "BoardPoint.h"
#include <iostream>

BoardPoint::BoardPoint(int x, int y) : x(x), y(y) {
}

BoardPoint BoardPoint::shiftLeft(int delta) const noexcept {
	return BoardPoint(x - delta, y);
}

BoardPoint BoardPoint::shiftLeft() const noexcept {
	return shiftLeft(1);
}

BoardPoint BoardPoint::shiftRight(int delta) const noexcept {
	return BoardPoint(x + delta, y);
}

BoardPoint BoardPoint::shiftRight() const noexcept {
	return shiftRight(1);
}

BoardPoint BoardPoint::shiftBottom(int delta) const noexcept {
	return BoardPoint(x, y + delta);
}

BoardPoint BoardPoint::shiftBottom() const noexcept {
	return shiftBottom(1);
}

BoardPoint BoardPoint::shiftTop(int delta) const noexcept {
	return BoardPoint(x, y - delta);
}

BoardPoint BoardPoint::shiftTop() const noexcept {
	return shiftTop(1);
}

int BoardPoint::getX() const noexcept {
	return this->x;
}

int BoardPoint::getY() const noexcept {
	return this->y;
}

bool BoardPoint::isOutOfBoard(int size) const noexcept {
	return x >= size || y >= size || x < 0 || y < 0;
}

void BoardPoint::print() const noexcept {
	std::cout << "[" << x << "," << y << "]" << "\n";
}

bool operator==(const BoardPoint& p1, const BoardPoint& p2) {
	return ((p1.getX() == p2.getX()) && (p1.getY() == p2.getY()));
}